### PR TITLE
Adding Email Subscription Component

### PR DESCRIPTION
**Pull request by tim-nguyen-cs**
Problem
=======
As a user,
I would like to subscribe to CruzHacks' newsletter
to receive timely updates about this years hackathon!

[Link to Asana Ticket](https://app.asana.com/0/1197738625416321/1197738625416326/f)

Solution
========
To create an email subscription component, we first create a responsive element based on the provided Figma design, which only includes an input field and submission button. The two elements are separated by a single right margin on `input` and wraps left-aligned using `flex-wrap: wrap`. The `input` component itself then resizes according to the width of the current viewports.

To handle user input, the `input`'s `onCapture` event is handled, which updates a local state variable corresponding to the user's current input. We also capture its `keyDown` event to submit their email whenever they input the "Enter" key. On the `button` component, we then submit their email when the button is clicked.

To create an asynchronous request, Axios is used to send an HTTP request to the designated Mailchimp endpoint and a feedback message is appropriately bound and displayed to the user using the `showFeedback` conditional.

Change summary:
---------------
* Added more accurate `Arrow` SVG and updated minimum breakpoint to include iPhone X 475c3fd0575617db4287fd9542c4f65d1387fa01
* Finished responsive layout e9ed21da4223269e0b97f8396e712bd1ace53067
* Handled user input  e31bb975dc125051302aa380fe688c7ef09e4c7e
* Configured Axios request 94c3d81be87ff24c139329ee9c02820c72cb6d9e

Steps to Verify:
----------------
1. Checkout branch using `git checkout feature/email-subscription`
2. Start local development server using `yarn start`
3. Open Chrome Developer Tools using `CMD/CTRL + SHIFT + I` to verify that email subscription component is responsive
4. Submit email in form to verify that appropriate response message is returned

### CODE REVIEW

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (tim-nguyen-cs) [inline src/components/EmailSubscription/EmailSubscription.scss:11]: I know we have designated fonts in `_fonts.scss` and that we were talking to Will about this, but what's the right way to import fonts from that file?

Comment (ilopezro) [reply src/components/EmailSubscription/EmailSubscription.scss:11]: That works! Take a look at how it's done in the `_fonts.scss` file. We defined a constant at the top and use that constant throughout the file.

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (ilopezro) [inline src/components/EmailSubscription/EmailSubscription.scss:39]: Let's use the BEM Naming Conventions for these. I looked at the TS file and it looks like the nested props all live under `EmailSubscription`. Let's rename them to `EmailSubscription__inputButton`, etc. We can keep the nestings here and just make use of the ampersand convention that SCSS provides. `&__inputButton`

Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:39]: I'm not sure what you mean by using the ampersand convention, because each subcomponent doesn't inherit the default `EmailSubscription` class, but take a look at the revisions that I made, and let me know if there's anything else that I need to do.

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (ilopezro) [inline src/components/EmailSubscription/EmailSubscription.scss:10]:
```suggestion
  .&__searchBox {
```

Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: Ohh, that's what you meant!

Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: @ilopezro - any recommendations for circumventing the `selector-class-pattern`  StyleLint error: `Classes must be in BEM form like `MyComponent__element--variant``

If we include the ampersand in the class name, it doesn't match the convention.

Comment (ilopezro) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: I'll take a look at it!

File: .stylelintrc.js
Comment (ilopezro) [inline .stylelintrc.js:12]: nesting depth is 2, max compound selector should go back to 1. We want to enforce these rules and not change them to our convenience.

File: .prettierrc.js
Comment (ilopezro) [inline .prettierrc.js:4]: Same here. Please change these back.

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (ilopezro) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: Yes, so you have to 
```suggestion
  &__searchBox {
  }
  &__inputButton {
  }
  &__feedbackButton {
  }
```

File: .prettierrc.js
Comment (tim-nguyen-cs) [reply .prettierrc.js:4]: Ahh, this was partially my fault - I had default line endings as `CRLF` instead, so a bunch of errors were showing up.

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: But that's not how modifying classes work with ampersands... ampersands are only applicable if that element also inherits its respective parent class. Taking a look at the screenshot below, we can see the `input` and `button` elements do not inherit their parent `EmailSubscription` class, so the ampersands will not apply correctly (also tested).

Can you clarify you perspective?

image

Comment (ilopezro) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: Sounds good. Then let's move away from nesting if they don't inherit parent props. Instead of it all being under the `.EmailComponent` tag let's put them outside

Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: A little confused by what you said there - are you talking about removing the `EmailSubscription` class/wrapper-element?

Comment (ilopezro) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: By having everything nested within the `.EmailComponent` class, we are assuming that it's children inherit all of their props. Let's move whatever doesn't belong inside `.EmailComponent` outside of that class. So we can see a structure like 
```scss
.EmailComponet{...props}
.EmailComponent__inputButton {...props} 
```

Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: Okay, so just to understand you correctly - do you want me to extrapolate `searchBox` and `inputButton` into their own components? Wouldn't that be too verbose of an extrapolation - they're just HTML components.

Comment (ilopezro) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: https://github.com/CruzHacks/cruzhacks-2021-website/blob/8c0f58a4e70459d732df92953cf211e1e620c22c/src/components/EmailSubscription/EmailSubscription.scss#L4-L41

Comment (ilopezro) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: This is what I was doing before you suggested [nesting them](https://github.com/CruzHacks/cruzhacks-2021-website/pull/8#discussion_r506146642). So if they do not inherit any of the props, just have it go outside of the `.EmailSubscription` component and have it be it's own component. This will help with readability.

Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: Sorry, I misunderstood you with my last comment. It's not too verbose and works alright, but I wouldn't say it's `its own component` - it's just not nested, which is different in my perspective.

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (w-gao) [inline src/components/EmailSubscription/EmailSubscription.scss:10]: I'm seeing this border on focus. Should be able to get rid of it with something like

```scss
&:focus {
  outline: none;
}
```
image

**Review by w-gao — COMMENTED** [review]
Overall, this looks great! You can probably constrain the width from the Hero view since it looks pretty large. I don't think there's any styling applied to the EmailSubscription component right now.

**Review by ilopezro — CHANGES_REQUESTED** [review]
Also, please fix this issue happening on your branch. I am not 100% sure what is happening or why it's happening, but the components are stacking on a normal laptop view, which shouldn't be happening at all.

Screen Shot 2020-10-26 at 6 05 44 PM

**Comment by mocklee** [comment]
@ilopezro the early flex-wrap is likely because of the postcard... its responsiveness might need some slight reworking to gradually decrease postcard size, so it's not constant width. I'll rework those responsiveness updates once Tim's branch is finalized

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (mocklee) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: I wouldn't remove input outlines; they're an accessibility feature https://www.a11yproject.com/posts/2013-01-25-never-remove-css-outlines/

File: package.json
Comment (ilopezro) [inline package.json:15]: There seems to be an issue with the merge with main. Please remove these `@types` from the dependencies and keep them in your `devDependencies`. I rebased main with your branch locally it fixed the issue! Whenever you're fixing merge conflicts, please be aware of what the changes are.

File: src/components/EmailSubscription/index.view.tsx
Comment (mocklee) [inline src/components/EmailSubscription/index.view.tsx:25]: Let's try to contain implementation constants (like API endpoints) as environment secrets when possible

**Review by mocklee — CHANGES_REQUESTED** [review]
backtracking a bit on the responsiveness issue @ilopezro raised— it seems that the early `flex-wrap` is indeed mostly being caused by the width of the email input; that component takes up the width of its entire parent. Try adding a `max-width` to the email input container, rather than `width: 100%`! 

Maybe the `max-width` should go on the parent container instead, it will take some tweaking to converge on properties that work.

one more gripe— there isn't any immediate user feedback when the send button is pressed. Maybe change the state of the input placeholder on submission w/ a loading message before vanishing the email component completely. Also, consider only vanishing the component on successful submission, and leaving an error message in the placeholder text when submit fails. The 2020 email component has a good example of this kind of UX (although the input is never locked)!

File: src/components/EmailSubscription/index.view.tsx
Comment (mocklee) [inline src/components/EmailSubscription/index.view.tsx:35]: this error message may be too long— we will need to think about wrapping this text on mobile

**Comment by ilopezro** [comment]
To piggy back @mocklee 's comments, the component itself looks really good. I'd suggest now adding the styling to the `Hero` view to make it all come together!

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:10]: The Figma design calls for no outline on the input box. What's the call here?

File: src/components/EmailSubscription/index.view.tsx
Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/index.view.tsx:25]: Can you clarify again how this is done or point me toward resources that help?

Comment (ilopezro) [reply src/components/EmailSubscription/index.view.tsx:25]: Working on this. Will get back to you ASAP!

File: src/stylesheets/_responsiveness.scss
Comment (ilopezro) [inline src/stylesheets/_responsiveness.scss:3]: I'd stray away from changing these constants. Please, if any, try and adjust the component using a `min-width` or a `max-width`.

File: src/components/EmailSubscription/index.view.tsx
Comment (ilopezro) [inline src/components/EmailSubscription/index.view.tsx:29]: I'd create a variable for this and instead of just have the variable value in here. 

```tsx
const subscribeEndpoint = process.env.REACT_APP_SUBSCRIBE_ENDPOINT || ""
...
axios.post(subscribeEndpoint)...
```

This will allow for cleaner and more readable code.

File: src/stylesheets/_responsiveness.scss
Comment (tim-nguyen-cs) [reply src/stylesheets/_responsiveness.scss:3]: I mentioned this above, but I changed this to accommodate for smaller phone resolutions, such as the iPhone X - I can change this back if need be.

I'm also confused about what you mean by your second comment. I've checked the responsive of the component itself, and I think it looks fine. Let me know if there's something specific (per screen resolution) you want me to take a look at!

(In short, `max-width` on `EmailSubscription` can work, but it might not be necessary.)

File: src/components/EmailSubscription/EmailSubscription.scss
Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/EmailSubscription.scss:9]: @mocklee Bump!

File: src/stylesheets/_responsiveness.scss
Comment (ilopezro) [reply src/stylesheets/_responsiveness.scss:3]: When you import the component here: 
https://github.com/CruzHacks/cruzhacks-2021-website/blob/9b61e7d46b04f92d808f3bea3949577917232fb8/src/views/Hero/index.view.tsx#L27-L32
You should wrap this within its own container and style that to fit within the larger view.

Comment (tim-nguyen-cs) [reply src/stylesheets/_responsiveness.scss:3]: Little confused about the difference, but I implemented it regardless! Let me know if we should change back the smallest resolution also, but take a look at it!

Comment (ilopezro) [reply src/stylesheets/_responsiveness.scss:3]: Because remember that each component is styled by itself in the `components` directory, but we need to integrate the component to each view and that's where we fix widths and all that good stuff.

Comment (ilopezro) [reply src/stylesheets/_responsiveness.scss:3]: As to your earlier point about changing the dimension of this, then I think you should be ok! I'm not sure how we came up with the breakpoints last year so this change is ok!

File: src/components/EmailSubscription/index.view.tsx
Comment (ilopezro) [reply src/components/EmailSubscription/index.view.tsx:42]: Bump in this @tim-nguyen-cs

File: src/components/EmailSubscription/index.view.tsx
Comment (ilopezro) [inline src/components/EmailSubscription/index.view.tsx:3]: These absolute paths looking clean 😤

**Comment by ilopezro** [comment]
I'm being super nit picky, but let's try and work on the responsiveness of the component a bit more? I'm wondering why the SVG looks lopsided in this view (iPhone X). Also, let's try and have the component take up the entire width of that whole div. 

Screen Shot 2020-10-28 at 10 55 20 PM

File: src/components/EmailSubscription/index.view.tsx
Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/index.view.tsx:42]: Take a look at how I hard-coded the widths on my current revision (pushed) and current iteration (not yet pushed), and let me know what else we should do:
```
  &__feedbackMessage {
    @include typography.description;
    color: var(--wedgewood);
    width: 350px;
    margin-top: var(--extra-small-spacer);

    @include responsiveness.respond(mission) {
      width: min(48vw, 300px);
    }
  }
```

In short, it wraps relative to the search box.

File: src/components/EmailSubscription/index.view.tsx
Comment (tim-nguyen-cs) [reply src/components/EmailSubscription/index.view.tsx:3]: Resolving as a parenthetical comment!

**Comment by tim-nguyen-cs** [comment]
@ilopezro Take a look at the newest revision! For some reason, I couldn't also get that error you showed to replicate on my end - here's what I currently see:

image

Let me know what else we can do!

**Comment by mocklee** [comment]
@tim-nguyen-cs Looks like the flex elements aren't justified towards the center of their rows— try adding `justify-content: center` on the parent flex containers https://css-tricks.com/almanac/properties/j/justify-content/#demo

**Comment by tim-nguyen-cs** [comment]
@mocklee I added `margin: auto` on the parent container to center align that one element only. If we add justify-content: center` on the entire "text column" container, it'll center align all other elements (unless this is something we want). Let me know if this works (not also that the screenshot above was not the most recent iteration that is currently pushed - that was just to showcase that I couldn't replicate Isai's error).

**Review by mocklee — APPROVED** [review]

**Review by ilopezro — CHANGES_REQUESTED** [review]
This LGTM, but let's hold off merging this until #17 gets merge into this branch. I'm trying to work something out with @mocklee

**Review by ilopezro — APPROVED** [review]
🚢

File: src/components/EmailSubscription/index.view.tsx
Comment (ilopezro) [inline src/components/EmailSubscription/index.view.tsx:32]: It's a very simple fix! Just add our email here. No need for a hyperlink. 
```suggestion
          "Oh no! We've got an error— please try your request again & contact us at dev@cruzhacks.com if this persists!"
```

**Pull request merged**
